### ISSUE TITLE
Infinite loop in TLS streams in Java 11

### ISSUE THREAD

**Issue by jroper** [akka/akka#28092]
I'm seeing, when I make a request in akka-http using https, the TLS actor appears to go in an infinite loop. I'm raising in Akka because the problem appears to be in Akka streams TLS support, not in akka-http.

It's not every request, health check requests from kubernetes don't appear to trigger it. Each of the requests that I make with curl trigger it, until the thread pool is exhausted and the application shuts down. Note that the requests that I'm making do succeed.

You can see the the two threads here reported in top:

```
    108 1001      20   0 4206272 477328  21268 R 98.7  3.1   3:37.06 management-api-                                                                         
    107 1001      20   0 4206272 477328  21268 R 97.4  3.1   3:53.86 management-api-  
```

These same threads stay in an infinite loop, it never switches to other threads, so it's likely inside a single receive invocation. Each additional request I make causes a new thread to start spinning at or near 100% cpu.

Here's the stack of one of those threads in a thread dump (note it's definitely the same thread, nid=0x6b which is decimal 107, shown above):

```
"management-api-akka.actor.default-dispatcher-26" #50 prio=5 os_prio=0 cpu=240772.98ms elapsed=512.59s tid=0x00007f673005f000 nid=0x6b runnable  [0x00007f66ea8f4000]
   java.lang.Thread.State: RUNNABLE
	at akka.stream.impl.TransferState.isExecutable(Transfer.scala:101)
	at akka.stream.impl.TransferState.isExecutable$(Transfer.scala:101)
	at akka.stream.impl.TransferState$$anon$6.isExecutable(Transfer.scala:108)
	at akka.stream.impl.Pump.pump(Transfer.scala:201)
	at akka.stream.impl.Pump.pump$(Transfer.scala:200)
	at akka.stream.impl.io.TLSActor.pump(TLSActor.scala:50)
	at akka.stream.impl.BatchingInputBuffer.enqueueInputElement(ActorProcessor.scala:96)
	at akka.stream.impl.BatchingInputBuffer$$anonfun$upstreamRunning$1.applyOrElse(ActorProcessor.scala:147)
	at scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:38)
	at akka.stream.impl.SubReceive.apply(Transfer.scala:18)
	at akka.stream.impl.FanIn$InputBunch$$anonfun$subreceive$1.applyOrElse(FanIn.scala:243)
	at scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:38)
	at akka.stream.impl.SubReceive.apply(Transfer.scala:18)
	at akka.stream.impl.SubReceive.apply(Transfer.scala:14)
	at scala.PartialFunction.applyOrElse(PartialFunction.scala:127)
	at scala.PartialFunction.applyOrElse$(PartialFunction.scala:126)
	at akka.stream.impl.SubReceive.applyOrElse(Transfer.scala:14)
	at scala.PartialFunction$OrElse.applyOrElse(PartialFunction.scala:175)
	at akka.actor.Actor.aroundReceive(Actor.scala:532)
	at akka.actor.Actor.aroundReceive$(Actor.scala:530)
	at akka.stream.impl.io.TLSActor.aroundReceive(TLSActor.scala:50)
	at akka.actor.ActorCell.receiveMessage(ActorCell.scala:573)
	at akka.actor.ActorCell.invoke(ActorCell.scala:543)
	at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:269)
	at akka.dispatch.Mailbox.run(Mailbox.scala:230)
	at akka.dispatch.Mailbox.exec(Mailbox.scala:242)
	at java.util.concurrent.ForkJoinTask.doExec(java.base@11.0.4/ForkJoinTask.java:290)
	at java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(java.base@11.0.4/ForkJoinPool.java:1020)
	at java.util.concurrent.ForkJoinPool.scan(java.base@11.0.4/ForkJoinPool.java:1656)
	at java.util.concurrent.ForkJoinPool.runWorker(java.base@11.0.4/ForkJoinPool.java:1594)
	at java.util.concurrent.ForkJoinWorkerThread.run(java.base@11.0.4/ForkJoinWorkerThread.java:177)
```

Same thread at another point in time in case it's interesting:

```
	at sun.security.ssl.SSLEngineImpl.getHandshakeStatus(java.base@11.0.4/SSLEngineImpl.java:801)
	- locked <0x00000000ec62ec50> (a sun.security.ssl.SSLEngineImpl)
	at sun.security.ssl.SSLEngineImpl.writeRecord(java.base@11.0.4/SSLEngineImpl.java:249)
	at sun.security.ssl.SSLEngineImpl.wrap(java.base@11.0.4/SSLEngineImpl.java:136)
	- eliminated <0x00000000ec62ec50> (a sun.security.ssl.SSLEngineImpl)
	at sun.security.ssl.SSLEngineImpl.wrap(java.base@11.0.4/SSLEngineImpl.java:116)
	- locked <0x00000000ec62ec50> (a sun.security.ssl.SSLEngineImpl)
	at javax.net.ssl.SSLEngine.wrap(java.base@11.0.4/SSLEngine.java:479)
	at akka.stream.impl.io.TLSActor.doWrap(TLSActor.scala:363)
	at akka.stream.impl.io.TLSActor.$anonfun$flushingOutbound$1(TLSActor.scala:239)
	at akka.stream.impl.io.TLSActor$$Lambda$1530/0x0000000100a96840.apply$mcV$sp(Unknown Source)
	at akka.stream.impl.Pump.pump(Transfer.scala:202)
	at akka.stream.impl.Pump.pump$(Transfer.scala:200)
	at akka.stream.impl.io.TLSActor.pump(TLSActor.scala:50)
```

Some environment stats:

akka 2.6.0-M8
akka-http 10.1.10
OpenJDK Runtime Environment AdoptOpenJDK (build 11.0.4+11)
Linux 4.14.138 running in a container on GKE

**Comment by jroper** [comment]
I've just tried with Java 8:

OpenJDK Runtime Environment (AdoptOpenJDK)(build 1.8.0_222-b10)

And the problem doesn't appear to be there. So, has most likely got something to do with Java 11.

**Comment by jroper** [comment]
This SFO post reports an infinite loop problem in the HTTP client, which may or may not be related, and it implies that the issue may be in the JDK, specifically with TLS 1.3 support:

https://stackoverflow.com/questions/54485755/java-11-httpclient-leads-to-endless-ssl-loop

I'll try disabling TLS 1.3 and see if that solves it.

**Comment by jroper** [comment]
Ok I can confirm it only happens with TLS 1.3, which explains why my curl requests were triggering it, but not other requests. Disabling TLS 1.3 fixes it. I'm going to do one more test, and that is see if it's still a problem in JDK13. If not, then we can conclude that this is a JDK issue and close this bug.

**Comment by jroper** [comment]
It doesn't appear to be a problem with JDK13 using TLS 1.3. There is one slightly different thing about this test, there are no adoptopenjdk 13 docker image builds, so I used openjdk instead. The openjdk image didn't have top available. I confirmed that it wasn't an issue by making many requests, and finding that it didn't stop responding, and then I also took a thread dump and saw there were no threads in state reported above. Exact version that it didn't happen in:

```
openjdk version "13.0.1" 2019-10-15
OpenJDK Runtime Environment (build 13.0.1+9)
OpenJDK 64-Bit Server VM (build 13.0.1+9, mixed mode, sharing)
```

I'm going to close this issue because I think we can be confident that it's not an Akka issue, rather it's a JDK11 issue, and we should advise people to disable TLS 1.3.

**Comment by jroper** [comment]
Some further info, the original problem that I was trying to diagnose was why some requests made through the Google Cloud HTTP load balancers were really slow, and others not. To eliminate the Google load balancers as the cause, I connected directly, and that's when I encountered this issue, and I assumed it was the same thing. Turns out it's not, on JDK13, with TLS 1.3 enabled, where this issue doesn't exist, the problem still existed with Google load balancers. But, when I disabled TLS 1.3, the problem went away. So, that problem may either be in the JDK or Google's load balancers.

The upshot of this - never use TLS 1.3, especially not when you're on the JDK. It's not worth losing a day of your time to.
